# Ebay Car Sales Data Analyses

We will clean and analyse car sales data.

* dateCrawled - When this ad was first crawled. All field-values are taken from this date.
* name - Name of the car.
* seller - Whether the seller is private or a dealer.
* offerType - The type of listing
* price - The price on the ad to sell the car.
* abtest - Whether the listing is included in an A/B test.
* vehicleType - The vehicle Type.
* yearOfRegistration - The year in which the car was first registered.
* gearbox - The transmission type.
* powerPS - The power of the car in PS.
* model - The car model name.
* kilometer - How many kilometers the car has driven.
* monthOfRegistration - The month in which the car was first registered.
* fuelType - What type of fuel the car uses.
* brand - The brand of the car.
* notRepairedDamage - If the car has a damage which is not yet repaired.
* dateCreated - The date on which the eBay listing was created.
* nrOfPictures - The number of pictures in the ad.
* postalCode - The postal code for the location of the vehicle.
* lastSeenOnline - When the crawler saw this ad last online.

In [301]:
import numpy as np
import pandas as pd

In [302]:
autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [303]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [304]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [305]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [306]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest','vehicle_type', 'registration_year', 'gearbox', 'powerPS', 'model','odometer', 'registration_month', 'fuel_type', 'brand','unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code','last_seen']

In [307]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  powerPS   model   odometer  registration_month fuel_type  \
0    manuell      158  andere  150,000km                   3       lpg   
1  automatik      286     7er  150,000km                   6    benzin   
2    manuell      102    golf   70,000km                   7    benzin   
3  automatik       71  fortwo   70,000km                   6    benzin   
4    manuell        0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Changed the column names from camelCase to snake_case as it is prefered by python.

In [308]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-10 15:36:24  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox       powerPS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [309]:
autos['seller'].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [310]:
autos['offer_type'].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

* seller column has 49999 - private | 1 - commercial(gewerblich)
* offer_type has 49999 - Angebot | 1 - Gesuch
* odometer and price column needs to be cleaned and converted to int/float

In [311]:
autos[['odometer', 'price']].head()

odometer   price
0  150,000km  $5,000
1  150,000km  $8,500
2   70,000km  $8,990
3   70,000km  $4,350
4  150,000km  $1,350

In [312]:
autos['price'].isnull().sum()

0

In [313]:
autos['odometer'].isnull().sum()

0

No null value is present in odometer and price column

In [314]:
autos['odometer'] = autos['odometer'].str.replace("km", "")
autos['odometer'] = autos['odometer'].str.replace(",", "")

In [315]:
autos['odometer'] = autos['odometer'].astype(int)

In [316]:
autos['price'] = autos['price'].str.replace("$", "")
autos['price'] = autos['price'].str.replace(",", "")

In [317]:
autos['price'] = autos['price'].astype(int)

Changed the odometer and price column to int type

In [318]:
autos.rename({"odometer":"odometer_km"}, inplace = True, axis = 1)

Renamed the odometer column to odometer_km to show units

***Closer look on odometer and price column***

In [319]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [320]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [321]:
np.sum(autos["price"] == 0)

1421

In [322]:
autos[autos['price'] == autos['price'].max()]

date_crawled                         name  seller offer_type  \
39705  2016-03-22 14:58:27  Tausch_gegen_gleichwertiges  privat    Angebot   

          price   abtest vehicle_type  registration_year    gearbox  powerPS  \
39705  99999999  control    limousine               1999  automatik      224   

          model  odometer_km  registration_month fuel_type          brand  \
39705  s_klasse       150000                   9    benzin  mercedes_benz   

      unrepaired_damage           ad_created  nr_of_pictures  postal_code  \
39705               NaN  2016-03-22 00:00:00               0        73525   

                 last_seen  
39705  2016-04-06 05:15:30

In [323]:
autos['price'].sort_values(ascending=False)

39705    99999999
42221    27322222
39377    12345678
47598    12345678
27371    12345678
2897     11111111
24384    11111111
11137    10000000
47634     3890000
7814      1300000
22947     1234566
43049      999999
514        999999
37585      999990
36818      350000
14715      345000
34723      299000
35923      295000
12682      265000
47337      259000
38299      250000
37840      220000
40918      198000
43668      197000
28090      194000
20351      190000
17140      180000
11433      175000
32840      169999
18509      169000
           ...   
21171           0
24900           0
1220            0
31264           0
24871           0
2470            0
21256           0
29503           0
37021           0
21253           0
47861           0
29482           0
31336           0
35819           0
27020           0
4266            0
2481            0
15208           0
33620           0
12819           0
31332           0
35821           0
8438            0
43925           0
38832     

In [324]:
acceptable = (autos['price']<1000) | (autos['price']>=4800000)

In [325]:
autos[acceptable].size

227360

In [326]:
autos.drop(autos[acceptable].index, axis=0, inplace=True)

In [327]:
autos['price'].describe()

count    3.863200e+04
mean     7.498207e+03
std      2.535787e+04
min      1.000000e+03
25%      2.200000e+03
50%      4.350000e+03
75%      8.950000e+03
max      3.890000e+06
Name: price, dtype: float64

Removed the ad listing within price below 1000 and above and equal to 4800000 (price of most expensive car as per 2016) because the price listed was unreal/too high or too low.

In [328]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  powerPS   model  odometer_km  registration_month fuel_type  \
0    manuell      158  andere       150000                   3       lpg   
1  automatik      286     7er       150000                   6    benzin   
2    manuell      102    golf        70000                   7    benzin   
3  automatik       71  fortwo        70000                   6    benzin   
4    manuell        0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [329]:
autos[['date_crawled', 'ad_created', 'last_seen']].head()

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [330]:
autos['date_crawled'].str[:10].sort_index().head()

0    2016-03-26
1    2016-04-04
2    2016-03-26
3    2016-03-12
4    2016-04-01
Name: date_crawled, dtype: object

In [331]:
autos['date_crawled'].str[8:10].value_counts(normalize=True, dropna=False).sort_index()

01    0.034609
02    0.036291
03    0.039139
04    0.036912
05    0.038906
06    0.017136
07    0.036628
08    0.032615
09    0.032460
10    0.033314
11    0.032797
12    0.037378
13    0.015997
14    0.036628
15    0.033625
16    0.029069
17    0.030493
18    0.012839
19    0.035126
20    0.038155
21    0.037301
22    0.032538
23    0.032201
24    0.029017
25    0.030519
26    0.033107
27    0.031399
28    0.035359
29    0.033987
30    0.033055
31    0.031399
Name: date_crawled, dtype: float64

In [332]:
autos['ad_created'].str[8:10].value_counts(normalize=True, dropna=False).sort_index()

01    0.034583
02    0.036136
03    0.040329
04    0.038854
05    0.035075
06    0.018508
07    0.035178
08    0.032564
09    0.032693
10    0.033081
11    0.033107
12    0.037171
13    0.017680
14    0.035049
15    0.033444
16    0.029690
17    0.030208
18    0.013331
19    0.034143
20    0.038284
21    0.037637
22    0.032382
23    0.032046
24    0.029069
25    0.030752
26    0.033237
27    0.031451
28    0.035644
29    0.034194
30    0.032926
31    0.031554
Name: ad_created, dtype: float64

In [333]:
autos['last_seen'].str[8:10].value_counts(normalize=True, dropna=False).sort_index()

01    0.023193
02    0.024928
03    0.024436
04    0.023374
05    0.132248
06    0.238248
07    0.144517
08    0.006238
09    0.008905
10    0.009811
11    0.011726
12    0.022184
13    0.008387
14    0.011985
15    0.014988
16    0.015454
17    0.026377
18    0.007377
19    0.014599
20    0.019802
21    0.019673
22    0.020786
23    0.017913
24    0.018534
25    0.017757
26    0.016075
27    0.014082
28    0.019440
29    0.020786
30    0.023452
31    0.022727
Name: last_seen, dtype: float64

In [334]:
autos['registration_year'].describe()

count    38632.000000
mean      2005.678194
std         86.678597
min       1000.000000
25%       2001.000000
50%       2005.000000
75%       2009.000000
max       9999.000000
Name: registration_year, dtype: float64

In [335]:
autos['registration_year'].sort_values()

22316    1000
49283    1001
21416    1927
22101    1929
11246    1931
2221     1934
2573     1934
39725    1937
21421    1937
23804    1937
26607    1937
26103    1938
24855    1939
13963    1941
25792    1941
11585    1943
11047    1948
14020    1950
35921    1951
44406    1951
23372    1952
35453    1953
40765    1954
25556    1954
24515    1955
36794    1955
7294     1956
26570    1956
23483    1956
19914    1956
         ... 
34111    2018
7137     2018
43577    2018
5164     2018
25799    2018
41042    2018
19373    2018
46986    2018
43589    2018
48065    2018
41084    2018
33999    2018
3848     2018
34099    2018
41111    2018
34107    2018
41045    2018
47016    2018
49185    2019
27578    2800
4549     4100
453      4500
22799    5000
49153    5000
27618    5911
8360     6200
25003    8888
49910    9000
33950    9999
38076    9999
Name: registration_year, Length: 38632, dtype: int64

In [336]:
autos[autos['registration_year'] == 2016]

date_crawled                                               name  \
76     2016-03-22 14:52:57                         BMW_318i_neustes_Model_0Km   
135    2016-03-12 11:00:10  Opel_Meriva_B_Panoramadach__Sitz__und_Lenkradh...   
256    2016-04-03 20:50:38                     Passat_1.9TDI_4Motion_Highline   
295    2016-03-28 03:36:22                                     Privat_anbiter   
307    2016-03-15 22:50:48                                       Giessen_ford   
374    2016-03-15 18:42:52                          Polo_9N_1_9_Tdi_Gti_Optik   
479    2016-03-08 07:55:41                              Renault_R4_TL_Savanne   
495    2016-03-17 18:45:02                                       BMW_1er_116i   
507    2016-03-08 13:45:53                                      Citroën_Xsara   
520    2016-03-07 00:52:52               Gut_gepflegter_Golf_GTI_zu_verkaufen   
574    2016-03-27 18:39:33                                   VW_vento_2_8_vr6   
696    2016-03-08 03:55:14                           Audi_a6_avant_2.5_tdi_v6   
1021   2016-03-25 08:57:19                     BMW_316i_Leder_Voll_FAHRBETEIT   
1133   2016-03-14 16:47:42                 Fiat_Panda_HP_100_Klimaautomatic!!   
1153   2016-03-17 23:53:33                    Audi_A3_Sportback_2.0TDI_S_Line   
1306   2016-04-01 20:37:43           Peugeot_607_HDI_2.2_mit_gruener_Plakette   
1326   2016-03-21 19:47:32          Opel_Meriva_A_in_einem_gepflegten_Zustand   
1474   2016-04-01 22:37:14         Golf_Vr6_Turbo_HGP_Turbo_Alles_Eingetragen   
1555   2016-03-17 01:00:21  Audi_Cabriolet____A5___NAVI_.XEN_._ALCANTARA.S...   
1562   2016-03-19 20:54:16  RENAULT_TWINGO___ERST_3_MONATE_ALT_UND_2000_KM...   
1593   2016-03-27 19:50:57                         BMW_520I_AUTOMATIKGETRIEBE   
1632   2016-03-29 23:48:06                                 Fiat_Idea_Mini_Van   
1741   2016-03-24 23:57:44                           Vw_Golf_4_2.0__121000_km   
1866   2016-03-07 17:42:12  MICROCAR_M.GO_Dynamic_Mopedauto_Leichtkraftfah...   
1956   2016-03-07 15:46:32                          Volkswagen_Passat_1_9_tdi   
1988   2016-03-17 20:36:21                      Polo_9n3_1.6l_105PS_Automatik   
2278   2016-03-15 18:53:59                           Bmw_318_Coupe_M_Original   
2280   2016-03-31 17:51:37                                        Ford_fiesta   
2446   2016-03-24 22:51:23                                    Golf_5__TDI_1.9   
2534   2016-03-07 09:57:07                                      Golf_1_Cabrio   
...                    ...                                                ...   
47848  2016-03-10 18:43:09       Alfa_romeo_zu_verkaufen_Liebhaber_Auto____:_   
47850  2016-03-26 22:37:33                               verkaufe_skoda_fabia   
47935  2016-03-08 10:38:12                            Bmw_e39_in_guten_zutant   
47982  2016-04-04 22:46:28                  Renault_Espace_2.0_136_PS_1998ccm   
48021  2016-03-20 22:52:25                           Golf_4_1_4_tuev_bis_6/16   
48058  2016-04-03 18:56:46  BMW_120D_Coupe_NAVI/_KEYLESS/LEDER/_TEMPOMAT/_PDC   
48062  2016-03-27 14:46:50                               Fiat_Scudo_1_6benzin   
48082  2016-04-01 21:55:53             Audi_A6_Avant_2.5_TDI_BJ_2003_Facelift   
48198  2016-04-04 11:57:12                            VW_Passat_3bg_tdi_131ps   
48254  2016-03-17 15:51:12                                  Ford_Focus_2_Tdci   
48275  2016-03-08 19:37:00                           FIAT_PUNTO_188_..._60_PS   
48317  2016-03-06 02:36:25  e36_325i_coupe_neulack_bilder_folgen_bbs_rf_st...   
48464  2016-03-17 15:37:39                 A_Klasse_170_CDI_Klima_sitzheizung   
48514  2016-03-28 20:49:59       Renault_Senic_Megane_16V_nur_34500km_TÜV_neu   
48752  2016-03-05 15:49:03               Opel_Astra_1_4T_MAHAGONIBRAUN_1.HAND   
48768  2016-04-03 18:54:34                       Golf_4_mit_Anhaengerkupplung   
48828  2016-03-24 22:51:18  Grand_Cherokee_ZJ__5_2l__neue_Stossdaempfer__a...   
48866  2016-03-17 09:51:16                   

In [337]:
reg_year_range = (autos['registration_year']<1927) | (autos['registration_year']>2016)

In [338]:
np.sum(reg_year_range)

1420

In [339]:
reg_year_range.index

Int64Index([    0,     1,     2,     3,     4,     5,     7,    12,    13,
               14,
            ...
            49988, 49990, 49992, 49993, 49994, 49995, 49996, 49997, 49998,
            49999],
           dtype='int64', length=38632)

In [340]:
autos.drop(autos[reg_year_range].index, axis=0, inplace=True)

In [341]:
autos['registration_year'].describe()

count    37212.000000
mean      2003.967618
std          7.021028
min       1927.000000
25%       2001.000000
50%       2005.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

Droped the registration year values which were impossibly old or had a registration year of the future which is again impossible.

In [342]:
(autos['registration_year'].value_counts(normalize=True).sort_index())*100
# multiplied by 100 to show percentages

1927    0.002687
1929    0.002687
1931    0.002687
1934    0.005375
1937    0.010749
1938    0.002687
1939    0.002687
1941    0.005375
1943    0.002687
1948    0.002687
1950    0.002687
1951    0.005375
1952    0.002687
1953    0.002687
1954    0.005375
1955    0.005375
1956    0.010749
1957    0.005375
1958    0.008062
1959    0.016124
1960    0.045684
1961    0.016124
1962    0.010749
1963    0.021498
1964    0.024186
1965    0.045684
1966    0.056433
1967    0.067183
1968    0.069870
1969    0.048371
          ...   
1987    0.137053
1988    0.284854
1989    0.330539
1990    0.507901
1991    0.556272
1992    0.588520
1993    0.585833
1994    0.728260
1995    1.190476
1996    1.456519
1997    2.114909
1998    3.447813
1999    4.635601
2000    5.377298
2001    5.549285
2002    5.737397
2003    6.656455
2004    7.008492
2005    7.484145
2006    7.126733
2007    6.070622
2008    5.922821
2009    5.581533
2010    4.256691
2011    4.345372
2012    3.509621
2013    2.136408
2014    1.7440

We can observe that from 1927 to 1994 very few cars were registered but after 1994 till 2007 there is a rapid increase in the number of cars registered again after 2007 there is gradual decrease till 2016.

In [343]:
autos['brand'].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'audi', 'renault', 'sonstige_autos', 'mazda', 'porsche', 'mini',
       'mercedes_benz', 'seat', 'toyota', 'opel', 'dacia', 'jeep', 'saab',
       'volvo', 'nissan', 'jaguar', 'skoda', 'subaru', 'fiat',
       'mitsubishi', 'chevrolet', 'hyundai', 'honda', 'kia', 'citroen',
       'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'rover',
       'daihatsu', 'daewoo', 'lancia', 'lada'], dtype=object)

In [344]:
brands = ((autos['brand'].value_counts(normalize=True))*100).head(6)

In [345]:
brands

volkswagen       21.081909
bmw              12.536279
mercedes_benz    11.157691
audi              9.757605
opel              8.905729
ford              5.871762
Name: brand, dtype: float64

In [346]:
popularity = brands

In [347]:
brands = brands.index

In [348]:
brands

Index(['volkswagen', 'bmw', 'mercedes_benz', 'audi', 'opel', 'ford'], dtype='object')

In [349]:
brand_mean_prices = {}
for brand in brands:
    mean_price = autos.loc[autos['brand']==brand, 'price'].mean()
    brand_mean_prices[brand] = mean_price

In [350]:
brand_mean_prices

{'audi': 10322.269347287249,
 'bmw': 9381.89174705252,
 'ford': 5786.703432494279,
 'mercedes_benz': 9302.614402697494,
 'opel': 4219.954737477368,
 'volkswagen': 6898.376545570427}

* Despite having only ~9% share audi's are much more expensive thus suitable for rich, that also explains the smaller share.
* Volkswagen has the majority share (ie- ~21%) and are moderately priced thus making them suitable for middle class families, that explains the large share.

In [351]:
brands

Index(['volkswagen', 'bmw', 'mercedes_benz', 'audi', 'opel', 'ford'], dtype='object')

In [352]:
brand_mean_mil = {}
for brand in brands:
    mean_mil = autos.loc[autos['brand'] == brand, 'odometer_km'].mean()
    brand_mean_mil[brand] = mean_mil

In [353]:
brand_mean_mil

{'audi': 127491.04929771413,
 'bmw': 132005.359056806,
 'ford': 119622.42562929062,
 'mercedes_benz': 130062.6204238921,
 'opel': 123952.92697646348,
 'volkswagen': 125771.82919056724}

Making a new dataframe for further analysis on brands.

In [354]:
bmp_series = pd.Series(brand_mean_prices)

In [355]:
bmm_series = pd.Series(brand_mean_mil)

In [356]:
pop_series = pd.Series(popularity)

In [357]:
brands_mean = pd.DataFrame(bmp_series, columns=['mean_price'])

In [358]:
brands_mean['mean_milage'] = bmm_series

In [359]:
brands_mean['popularity'] = pop_series

In [360]:
brands_mean

mean_price    mean_milage  popularity
audi           10322.269347  127491.049298    9.757605
bmw             9381.891747  132005.359057   12.536279
ford            5786.703432  119622.425629    5.871762
mercedes_benz   9302.614403  130062.620424   11.157691
opel            4219.954737  123952.926976    8.905729
volkswagen      6898.376546  125771.829191   21.081909

### Conclusions
* Ford had the worst milage and was priced more than the competetors thus was not very popular.
* BMW and mercedes_benz gave the best milage and therefore are quite expensive, but still were very popular.
* Volkswagen gave a balance between cost and milage.
* In general cars with higher milage were much more expensive that the cheaper ones.

*** Milage here means the distance that was travelled by the car in it's service time period, the cars with more milage meant that they were more satisfactory to the owner thus used for longer period before selling. ***

# Extras

* Further cleaning - converting the german words to english

In [362]:
autos['unrepaired_damage'] = autos['unrepaired_damage'].map({'nein':'No' , 'ja':'Yes'})

In [363]:
autos['unrepaired_damage'].unique()

array(['No', nan, 'Yes'], dtype=object)

In [365]:
autos['unrepaired_damage'].value_counts(dropna=False)

No     29707
NaN     5184
Yes     2321
Name: unrepaired_damage, dtype: int64

In [396]:
del_index = autos[autos['unrepaired_damage'].isnull()].index

In [400]:
autos.drop(del_index, axis=0, inplace=True)

In [401]:
damage_values = autos['unrepaired_damage'].unique()

In [403]:
damage_values

array(['No', 'Yes'], dtype=object)

In [423]:
damage_mean_price = {}
for damage_value in damage_values:
    mean_price = autos.loc[autos["unrepaired_damage"] == damage_value, "price"].mean()
    damage_mean_price[damage_value] = mean_price

In [424]:
damage_mean_price

{'No': 8274.669943111052, 'Yes': 4328.579922447221}

* It can be clearly seen that cars with no damage are almost **double** the value of cars with damage

In [432]:
autos['model'].value_counts(normalize=True)*100

golf                  7.896091
andere                7.558513
3er                   6.397891
passat                3.067130
a4                    2.980324
5er                   2.906379
c_klasse              2.896734
astra                 2.764918
polo                  2.629887
e_klasse              2.398405
corsa                 2.231224
a6                    2.121914
a3                    2.115484
transporter           1.755401
focus                 1.668596
1er                   1.530350
fortwo                1.433899
2_reihe               1.430684
a_klasse              1.263503
fiesta                1.224923
touran                1.150977
zafira                1.044882
3_reihe               0.948431
x_reihe               0.884131
mondeo                0.868056
octavia               0.842335
clio                  0.842335
vectra                0.832690
cooper                0.832690
twingo                0.813400
                        ...   
gl                    0.032150
kalos   

In [436]:
autos.loc[autos['model'] == 'golf', 'brand'].value_counts()

volkswagen    2456
Name: brand, dtype: int64

* The most common brand-model combination is **Volkswagen-Golf**.

In [441]:
autos['gearbox'].value_counts(dropna=False)

manuell      23313
automatik     8065
NaN            650
Name: gearbox, dtype: int64

In [449]:
mapping_dic = {'manuell':'manual', 'automatik':'automatic'}

In [450]:
autos['gearbox'] = autos['gearbox'].map(mapping_dic)

In [451]:
autos['gearbox'].unique()

array(['manual', 'automatic', nan], dtype=object)

In [452]:
autos['gearbox'].value_counts(dropna=False)

manual       23313
automatic     8065
NaN            650
Name: gearbox, dtype: int64

Translated few German words to English.